From 5adbd1944c9c45aa09a22bb301cb1442a8d7aa3f Mon Sep 17 00:00:00 2001
From: Prokhar Bhowal <Prokhar.Bhowal@amd.com>
Date: Fri, 13 May 2016 14:13:25 -0400
Subject: [PATCH 0486/1722] drm/amd/dal: Set CRTC_MASTER_UPDATE_MODE to 0

-set CRTC_MASTER_UPDATE_MODE to 0 for dce80, dce110, dce112

Signed-off-by: Prokhar Bhowal <Prokhar.Bhowal@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/dc/dce100/dce100_hw_sequencer.c | 10 +++++++++-
 drivers/gpu/drm/amd/dal/dc/dce112/dce112_hw_sequencer.c | 12 ++++++++++--
 drivers/gpu/drm/amd/dal/dc/dce80/dce80_hw_sequencer.c   | 12 ++++++++++--
 3 files changed, 29 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/dc/dce100/dce100_hw_sequencer.c b/drivers/gpu/drm/amd/dal/dc/dce100/dce100_hw_sequencer.c
index 496a8d0..1d73000 100644
--- a/drivers/gpu/drm/amd/dal/dc/dce100/dce100_hw_sequencer.c
+++ b/drivers/gpu/drm/amd/dal/dc/dce100/dce100_hw_sequencer.c
@@ -212,8 +212,16 @@ static bool dce100_enable_display_power_gating(
 	else
 		cntl = ASIC_PIPE_DISABLE;
 
-	if (!(power_gating == PIPE_GATING_CONTROL_INIT && controller_id != 0))
+        if (!(power_gating == PIPE_GATING_CONTROL_INIT && controller_id != 0)){
+
 		bp_result = dc_bios_enable_disp_power_gating(dcb, controller_id + 1, cntl);
+                /* Revert MASTER_UPDATE_MODE to 0 because bios sets it 2
+                * by default when command table is called
+                */
+                dm_write_reg(ctx,
+                        HW_REG_CRTC(mmMASTER_UPDATE_MODE, controller_id),
+                        0);
+        }
 
 	if (bp_result == BP_RESULT_OK)
 		return true;
diff --git a/drivers/gpu/drm/amd/dal/dc/dce112/dce112_hw_sequencer.c b/drivers/gpu/drm/amd/dal/dc/dce112/dce112_hw_sequencer.c
index 3354e93..c96c335 100644
--- a/drivers/gpu/drm/amd/dal/dc/dce112/dce112_hw_sequencer.c
+++ b/drivers/gpu/drm/amd/dal/dc/dce112/dce112_hw_sequencer.c
@@ -338,10 +338,18 @@ static bool dce112_enable_display_power_gating(
 	else
 		cntl = ASIC_PIPE_DISABLE;
 
-	if (power_gating != PIPE_GATING_CONTROL_INIT || controller_id == 0)
+        if (power_gating != PIPE_GATING_CONTROL_INIT || controller_id == 0){
+
 		bp_result = dc_bios_enable_disp_power_gating(
 						dcb, controller_id + 1, cntl);
-
+                /* Revert MASTER_UPDATE_MODE to 0 because bios sets it 2
+                 * by default when command table is called
+                 */
+                dm_write_reg(ctx,
+                        HW_REG_CRTC(mmCRTC_MASTER_UPDATE_MODE, controller_id),
+                        0);
+        }
+ 
 	if (power_gating != PIPE_GATING_CONTROL_ENABLE)
 		dce112_init_pte(ctx);
 
diff --git a/drivers/gpu/drm/amd/dal/dc/dce80/dce80_hw_sequencer.c b/drivers/gpu/drm/amd/dal/dc/dce80/dce80_hw_sequencer.c
index 92ed13c..d2be44b 100644
--- a/drivers/gpu/drm/amd/dal/dc/dce80/dce80_hw_sequencer.c
+++ b/drivers/gpu/drm/amd/dal/dc/dce80/dce80_hw_sequencer.c
@@ -194,10 +194,18 @@ static bool dce80_enable_display_power_gating(
 	else
 		cntl = ASIC_PIPE_DISABLE;
 
-	if (!(power_gating == PIPE_GATING_CONTROL_INIT && controller_id != 0))
+        if (!(power_gating == PIPE_GATING_CONTROL_INIT && controller_id != 0)){
+
 		bp_result = dc_bios_enable_disp_power_gating(
 						dcb, controller_id + 1, cntl);
-
+                /* Revert MASTER_UPDATE_MODE to 0 because bios sets it 2
+                 * by default when command table is called
+                 */
+                dm_write_reg(ctx,
+                        HW_REG_CRTC(mmMASTER_UPDATE_MODE, controller_id),
+                        0);
+        }
+ 
 	if (bp_result == BP_RESULT_OK)
 		return true;
 	else
-- 
2.7.4

